### PR TITLE
Fix/checksum

### PR DESCRIPTION
**Pull request by djarecka**
changes in `TaskBase.checksum` and `BaseSpec.hash`. 
My changes introduced in [this commit](https://github.com/nipype/pydra/commit/86467b7fa9335ec60bee051227188ffa73f798a6) that were fixing problem with no existing output_dir, didn't work for situations from `test_annotated_func` (test was xfailed..).

Main changes:
- all checksum can be calculated before running (also for state) 
- hash property is changed to a method that takes dictionary with state values as an optional argument (so can replace values for inputs that are part of the splitter)
- to calculate has of `_graph`, `nd.inputs.hash()` is calculated, so `TaskBase.checksum` diesn't have to deal with lazy inputs

### CODE REVIEW

File: pydra/engine/specs.py [hunk 1]
```diff
@@ -18,18 +18,19 @@ class SpecInfo:
...
             if field.name not in ["_graph"]
         }
+        if replace:
+            inp_dict.update(replace)
         inp_hash = self._hash(inp_dict)
         if hasattr(self, "_graph"):
-
-            graph_hash = [nd.checksum for nd in self._graph]
+            # calculating inputs.hash for every node inside the graph
+            graph_hash = [nd.inputs.hash() for nd in self._graph]
```
Comment (satra) [inline pydra/engine/specs.py]: if an internal node has splits, how would you know whether splits have changed or not.

File: pydra/engine/tests/test_task.py [hunk 2]
```diff
@@ -171,7 +162,7 @@ def test_shell_cmd(tmpdir):
     assert res.output.stdout == " ".join(cmd[1:]) + "\n"
 
     # separate command into exec + args
-    shelly = ShellCommandTask(executable=cmd[0], args=cmd[1:])
+    shelly = ShellCommandTask(name="test", executable=cmd[0], args=cmd[1:])
```
Comment (satra) [inline pydra/engine/tests/test_task.py]: should not be necessary.

**Comment by djarecka** [comment]
@satra - I added the changes after discussion. I've added some tests, including one for workflows that have the same inputs and nodes, but different spliter in one of the node.

Unfortunately, one test is still failing, looks like `inputs["_func"]` gives different hash for my osx and linux. will check it tomorrow, but let me know if you have any idea, looks like I had to introduce the error in this PR...

**Comment by satra** [comment]
@djarecka - since we are using cloudpickle it's quite possible that the dependencies are different between your setup and travis. may be worthwhile to test in a container with the same environment. but this is an issue we will need to address better. (perhaps in another PR).

**Comment by djarecka** [comment]
@satra - yes, `cp.dumps(func)` returns different values. And I didn't introduce it in this PR, I've just added this assert in the PR... I moved it to `test_task` that is xfailing and opened #87

File: pydra/engine/tests/test_task.py (hunk 2 above)
Comment (djarecka) [reply pydra/engine/tests/test_task.py]: was needed because `ShellCommandTask` didn't have default value for `name`. Added None

**Comment by satra** [comment]
i'm merging this together with a few things to be sorted later. also i would like to see if we can avoid the `deepcopy(self.graph_sorted)`

**Comment by djarecka** [comment]
@satra - see #92

**Pull request merged**